Document build/release pipelines (#205)
pace-neutrons/Horace: documentation/smg/10_build_release_process.md

When a pull request is opened or updated the code is built and tested of the merge of the PR branch and `master`.

1. Feature developed on a branch taken from the `master`
2. All tests likely to be affected by changes in Herbert and Horace must be run on developers machine
3. Create Herbert PR (if required) for merge into `master` branch
4. Create Horace PR (if required) - if a Herbert PR has been created too the Horace PR must not be merged until the Herbert PR has been merged
5. If the build / test against of the PR against the Release branch succeeds on all platforms the PR can be merge.

### Hot fixing

Patch releases should be tested and built through the same build pipeline as the initial production releases.

<img src="./images/10_git_hotfix_branches.png">

1. User reports issue
2. Bug report issue created in GitHub/Issues documenting how to reproduce
3. Fix developed on a branch taken from the *release* tag (`Rm_n_o`)
4. All tests likely to be affected by changes in Herbert and Horace must be run on developers machine
5. PR created for merge into Release branch (`rel_m_n_o`) *and* `master` branch

> 6. If the build / test against of the PR against the Release branch succeeds on all platforms, that build artifact can be released to the target platform as version `m.n.o+1`

Reviewer: Does the notation `m_n_o` here refer to semantic versioning i.e. major_minor_patch? Sorry if this has already been clearly established before. And does this imply that the only changes that will be made to a release branch will be patches?

If so then I agree :) I should probably adopt something similar for Euphonic

Author: Yes, I was thinking of `m.n.o` being semantic versioning; the idea that a "release" is `m.n.0`.

The only changes to that release will be patches/hotfixes as the next release will be 'm.n+1` and a new tag/branch.

By creating a tag on the `master` branch when the release it made it's then simple to go back and create the patch branch from that commit, which is a common ancestor for the release branch and current master.

None of that is written anywhere so I will add it.

# Build/Release pipeline

Build and test pipelines have been created in ANVIL for the Horace and Herbert source.

All users should be using a build that has been built, tested and packaged using the build script.

### Master

> The `master` branch should always be 'releasable' with all tests passing. Builds are run on all target platforms to created a `.zip`/`.tar.gz` release packages each night.

Reviewer 2: Typo `create`

- The Herbert `master` branch is built and a zip release build artifact with MATLAB-release and platform-specific binary is created.

- The Horace `master` branch is built and zip release build artifact with MATLAB-release and platform-specific binary created, which includes the last successful Herbert `master` build.

The Horace build artifact is a single deployable unit to a specific MATLAB/OS platform. This is equivalent to the artifact created by the legacy MATLAB create release process.

### Features

Feature development and bug fixes should be performed on branches and reviewed using a GitHub Pull Request (PR).

> When a pull request is opened or updated the code is built and tested of the merge of the PR branch and `master`.

Reviewer 2: typo `of the merge of`

1. Feature developed on a branch taken from the `master`
2. All tests likely to be affected by changes in Herbert and Horace must be run on developers machine
3. Create Herbert PR (if required) for merge into `master` branch
4. Create Horace PR (if required) - if a Herbert PR has been created too the Horace PR must not be merged until the Herbert PR has been merged
5. If the build / test against of the PR against the Release branch succeeds on all platforms the PR can be merge.

### Hot fixing

Patch releases should be tested and built through the same build pipeline as the initial production releases.

<img src="./images/10_git_hotfix_branches.png">

1. User reports issue
2. Bug report issue created in GitHub/Issues documenting how to reproduce
3. Fix developed on a branch (`xxx_branch`) taken from the *release* tag (`Rm_n_o`)
4. All tests likely to be affected by changes in Herbert and Horace must be run on developers machine
5. PR created for merge into Release branch (`rel_m_n_o`) *and* `master` branch. If this is the first release patch, the branch will need to be created

> 6. If the build / test against of the PR against the Release branch succeeds on all platforms, that build artifact can be released to the target platform as version `m.n.o+1`

Reviewer 2: typo `against of the PR against`

> If the hotfix is being done out-of-hours the PR can be merged by the developer WITHOUT review, but the branch should not be deleted and should be reviewed as the earliest opportunity.

Reviewer 2: I would add more to this. `merged by the developer WITHOUT review provided all tests pass`. Otherwise this is very risky.

The same caveats apply about the Horace build dependencies on Herbert.

Advantages:

- build tools are not required on the production hardware
- all builds are created using the same build tools
- users are using a known, traceable, validated release

Disadvantages:

> - slower; have to wait for full build and test pipeline to execute

Reviewer 2: This is only a problem if we do not get our quality assurance/control right. If we continue to improve testing at all levels as we encounter issues, this impact should lessen over time.

Reviewer 3: Mantid improves its assurance and control for more then 10 years. And on yesterday buziness meeting of our group number of our IS were complaining that Mantid is unreliable and crashing. 
Lets come rely on what we have and not on what happens if....

### Releases

> The Jenkins build artifacts are are not accessible outside of STFC. End users will be directed to GitHub to access packaged releases.

Reviewer 2: Mantid also uses [sourceforge](https://sourceforge.net/) for publishing releases. We could look into this. I think it is free for open-source software, not sure how it would be complicated by the matlab stuff.

Reviewer 3: we should be able to do sourceforge but currently is easier to continue with current package distribution which is downloadable from Horace web-site. Until Freddy is here, it is simple and reliable way of providing software to the users.

### Facility Deployment

Users participating in ISIS experiments use expensive neutron facility and need to be able to see and analyse their data as quickly as possible, ideally in the process of experiment and immediately after that. In addition to standard Horace scripts users sometimes write their own scripts, which places high demand on the software reliability. Sometimes, their custom scripts use standard software in the operational modes, which are difficult to predict. Often these script identify issues, which have not been anticipated during development so users need urgent fixing or modifications to the software. Sometimes, the issues can be caused by specific user's data, so can be easily reproduced only using these data. The changes fixing the user issue should occur immediately, as users should not lose a day or even an hour waiting until the whole build pipeline completes, produces all possible artefacts and get installed on the facility computers.  

> 

Reviewer 2: If we invest more time in developing tests around these scenarios it should lead to reducing these effects. We need to ensure we have a mechanism for capturing these issues (and possibly generalising them) or else we will forever be firefighting with hotfixes. We should make this a driver to improve quality control as opposed to a normal operating mode.

Author: ```
Often these script identify issues, which have not been anticipated during development so users need urgent fixing or modifications to the software. Sometimes, the issues can be caused by specific user's data, so can be easily reproduced only using these data. 
```
This is exactly the use-case for the "Hotfix" workflow described elsewhere in the document. The user script and data should be referenced/included in the issue ticket.

We should aim to minimise these failures though (increasingly comprehensive) testing, but agree that there will always be cases we haven't seen.

If your immediate fixes are rolled out without comprehensive testing there is a real danger of unintended bugs being introduced and Horace breaking for all the other users on the machine (as there is a common/shared code base). 

A decision must be reached on what "level" of support the PACE tooling has in production. What is an acceptable turn around time for a bug fix? In what situations could a workaround be identified more rapidly requiring a only small change to the users custom script, for example.

Reviewer 2: I am just worried we are indirectly creating a requirement to have a dev environment on every production instance.

Reviewer 3: until I have not fixed an issue during for at least three cycles in a row, the results of the investing in the tests will not be conclusinve so workflow is necessary.  At the moment I am fixing an average 3 issues per cycle. The waste majority is minor.

Reviewer 3: We are not creating the need for the enviroment -- Matlab is there anyway, and we are not chaning anything else. 

I am describing current level of the support. If somebody takes this task from me, he will decide on the best way of proceed. This way is proven to be successful for now. 

It also allows rapid introducing of the new features (equivalent of Mantid "nightly" edition)
As you know, this feature is in high demand on Mantid. Do not think we need it exactly Mantid way, as team and demands are order of magnitude smaller, but further morphing may come from this.

Reviewer 2: Nightlies are good strategies. But changes in our nightlies go through the full PR/Review/Build/Package process. There are no hotfixes applied in situ. Users may have to wait at least 24 hours for the new nightly build. Which is fundamentally different from what you are suggesting.

Author: The "nightly" (aka 'master') build can be manually triggered so the wait time would only be the combined Horace and Herbert build/test time, ~2.5hrs at the moment.

Author: Another thought triggered while summarising this discussion is how this model maps onto the SCARF deployment when that becomes available to users.

If a user encounters issues on iDAaaS they are likely to hit the same issue when running the analysis on bigger data on the SCARF cluster

Reviewer 3: looks like scarf deployment will me mainly compiled code deployment, so will not receive hotfixes.

Quick modifications to the software with only partial testing carries the risks of breaking the other parts of the software. Despite that, in reality, single crystal experiments on different machines rarely happen simultaneously. The combined probability of error in modified code and the situation when this error immediately and severely affects another high priority user can be considered low. If this happens, the possibility to revert changes as rapidly as they were introduced, is also very important. 

Because of the specific demands, the process of deploying the software on the ISIS facility differs from the  normal build-release process. 

Namely:

When a release is produced and placed to the github, the master branch is tagged with the release number. 

> The release tagged version is checked out in the common software area of the ISIS computing facility provided to users. The low level build artefacts i.e. C++ libraries are copied into appropriate places of this area from the Jenkins node, testing release on this system. 

Reviewer 4: Having the source on iDAaaS make me a bit uncomfortable. One of our main release platforms will never use a release generated by our release pipelines?

If what we need is a fast way to distribute a particular Git branch to users (i.e. without running the > hour long tests). There's a better way of achieving this:

We can create a new "hot-fix" pipeline, that does not run tests. Only builds and packages.

- Open a bug-fixing branch and implement your fix
- Run your branch through a "hot-fix" pipeline that just does build and package
	- This pipeline should take ~5 mins
- Retrieve the build artifact and install that on iDAaaS
- Open PR to merge your bug-fixing branch into the release branch
	- This will run all the tests - letting us know if we've broken something in ~ 1hr
	- This can be merged into the release branch
- We replace the hot-fix package with the release build artifact as soon as possible

The advantages of this are that:
- we (most of the time) have a tested and built version of Horace on iDAaaS
- we can rapidly update the version with a well-defined build
- any hot-fixed versions will have the commit-SHA appended to the version, so we get    traceability 
- we are no longer required to maintain a separate production branch
	- the only "maintained" branches are the release branches and master

> The release tagged version is checked out in the common software area of the ISIS computing facility provided to users. The low level build artefacts i.e. C++ libraries are copied into appropriate places of this area from the Jenkins node, testing release on this system. 

Author: This pipeline would, obviously, require that the developer has added new tests that are able to reproduce the bug (before fix) and so demonstrate that it is has been resolved when run on the developers machine (not on the iDAaaS deployment platform) after the fix is applied and “reasonable” execution of other test cases has been performed.

> The release tagged version is checked out in the common software area of the ISIS computing facility provided to users. The low level build artefacts i.e. C++ libraries are copied into appropriate places of this area from the Jenkins node, testing release on this system. 

Reviewer 3: iDaaaS users indeed do not have access to the packaged release unless packaged release is equal to the repository version. In addition, currently packaged release is p-coded on the private functions, which creates problems for bugfixing. I am slowly disabling this feature by-steals but never had explicit approvement for that. c-make pipeline fully disables this option, which it probably should. 

creating tests which reproduce bug may not be easy. This should be a purpose, but its not practical in all circumstances. 

Harry's pipeline looks attractive until you look at the way how real problem is discovered. 
Thanks god, we have a standard user tests so standard setup and workflow would work ok, but let's assume that something happened and a user reports error and you are invited to help him to process his data. 

1) you got users screen and enter user enviroment which may be complex (sometimes very complex e.g. Radu)
2) you look at the errors in user setup and its 50% chance that you fix the error on this stage. good. 
3) You are unlucky --it looks like user did everything right but Matlab code returns bunch of erros. 
4) at this stage error may be in user data, user setup or our code.  According to Harry, at this stage I need to go to my machine and start trying to recriate the user issue. 
5) Recriating thing which look like user setup are not easy. I would not spend half of a day to recriate Radu's setup especially if his experiment is running and after this time my setup will work. I need to check what is going on with user setup so open the debugger and go into the code. 
6) 2/3 it something with the data or user setup messing with our code. Problem fixed or moved to other place (data are corrupted)
7) with current ratio 1/3 is the bug in the code or something which looks like a bug in our code. 
8) Here is the judgement moment. Can I fix it without affecting other users? If I am unsure, I am telling user "sorry" and follow Harrys pipeline starting from recriating user setup and data on my enviroment. Very often, at this stage I can suggest user a workaround so we continue to work indepenently. 
9) Looks like simple bug fixing which not affect other users. I am fixing it, run short sequence of tests, probably related to the bug, and commit the change. User continues the experiment and I am creating PR to fully test the fixture and incorporate it into master.  all OK. 
10) I am not sure that the fix is not affecting other users but user is in hurry and it looks like 
safe way to go. There is normally low probablility that other users are using the same code at this particular moment of time so I do a fix. User works and I am following 8-9.
11) oops -- something went wrong at 9-10. Other users start complaining. (happened once or twice -- not that often for me to remember). 
there is high probability that the first user did everything he need at this stage. Noramlly at this stage I have open channel of communication with this user so see what is going on. Anyway, I am reverting the change I did (checkout branch), discuss issue with the first user, got some solution for him (including swithching him to development channel) and continue from point - 8. People not very happy, but see that their problems are dealt with so they are OK

Harrys pipeline constrains me to point 5 of the workflow and deprives me of 2/3 probablity of fixing issues without doing much work. with ~0.1 probabilty to create additional problems which are also fixable without big problem. 

Making correct and recriatable bug report is not an easy job which no one user likes to do and not everybody even able to produce. This link in Mantid is broken and I am extreamly happy that Horace is different.  Mantid is better now, (after 10 years of development) but the number of issues and unhappy users was huge on the way and some are still there.  As we have unique advantage (the core of the code is written in high level language) let's not loose it. 

I am completely against breaking working system for the sake of an abstract idea.

> If any errors are identified during user cycle, a bugfixing branch is created from the actual release. The changes, fixing user issue are committed to the github as soon as they are introduced. The changes become the contents of the bugfixing PR, which is reviewed, tested against all remaining code base and operating systems and eventually becomes the part of **master** like any other code changes. 

Reviewer 4: ```suggestion
If any errors are identified during user cycle, a bug-fixing branch is created from the relevant release. 
The changes, fixing the user's issue, are committed to GitHub as soon as they are completed. 
The changes become the contents of the bug-fixing PR, which is reviewed and tested against the existing codebase on all operating systems. 
Eventually, this PR will be merged into **master** like any other code change. 
```

> Such workflow allows rapid response to the requests of the users, who is carrying out their experiments, with only minor temporary decrease in the code reliability.

Reviewer 4: ```suggestion
Such a workflow allows rapid response to the requests of users who are carrying out their experiments,
with only a minor, temporary decrease in the code's reliability.
```

> The changes to low level performance code are considered high risk changes, so they are not introduced using this process. They are also not considered the critical changes as the code expected to run correctly without low level C++ code, may be with decreased performance. 

Reviewer 4: ```suggestion
The changes to low level performance code are considered high risk changes, so they are not introduced using this process. 
They are also not considered critical changes, as there should always be equivalent functionality implemented in Matlab. 
This may lead to decreased performance. 
```

4. Create Horace PR (if required) - if a Herbert PR has been created too the Horace PR must not be merged until the Herbert PR has been merged
5. If the build / test of the PR against the Release branch succeeds on all platforms, and the code has been reviewed, the PR can be merged.

### Releases

The Jenkins build artifacts are are not accessible outside of STFC. End users will be directed to GitHub to access packaged releases.

A 'deploy' pipeline will be created ([Horace #73](https://github.com/pace-neutrons/Horace/issues/73)).

This will:

- tag the `master` branch with the release number (`Rm_n_o`)
- copy the built artifact to GitHub

The release tag will be an anchor for any subsequently needed release branch and patch branches.

The version number is stored in a text file (`VERSION`) in the root of the Herbert and Horace source. This will follow [semantic versioning](https://semver.org/) and is used in the build process to set the version number in the MATLAB and compiled C++ components.

### Patch releases

> Patch releases will be made to releases branches to resolve specific bugs identified. These should be tested and built through the same build pipeline as the initial production releases.

Reviewer 4: ```suggestion
Patch releases will be made to release branches to resolve specific bugs identified. These should be tested and built through the same build pipeline as the initial production releases.
```

- slower; have to wait for full build and test pipeline to execute

### Hot-fix Releases

> If a particular fix is required quickly on a target system so that facility users to not lose critical machine time the full build and test process can be bypassed. 

Reviewer 4: ```suggestion
If a particular fix is required quickly on a target system, the full build and test process can be bypassed so that facility users do not lose critical machine time. 
```

Quick modifications to the software, with only partial testing, carry the risks of breaking other parts of the software, so this is an exceptional use-case. Non-time critical issues should be resolved through the standard Patch Release process.

> A "hot-fix" pipeline has been created in ANVIL that does not run tests and only builds and packages Herbert or Horace. This will complete is around 5 minutes as it is the test execution step of the standard PR pipeline that take in excess of an hour to run.

Reviewer 4: ```suggestion
A "hot-fix" pipeline has been created in ANVIL that does not run tests and only builds and packages Herbert or Horace. This will complete in around 5 minutes as it is the test execution step of the standard PR pipeline that take in excess of an hour to run.
```

Reviewer 4: We don't have this pipeline yet, we should open an issue for this.

Might also be worth splitting this line up into a few lines

1. User reports issue
2. Bug report issue created in GitHub/Issues documenting how to reproduce, and include any custom scripts or data that triggered the issue
3. Fix developed on a branch (`xxx_branch`) taken from the *release* tag (`Rm_n_o`)

> 4. New tests demonstrating the issue and it's resolution are developed and tests identified as being likely to be affected by the code code run on developer's machine

Reviewer 4: ```suggestion
4. New tests demonstrating the issue and its resolution are developed. Tests identified as being likely to be affected by the code changes run on developer's machine
```

> 5. Run the branch through the "hot-fix" build pipeline on Jenkins that *just* executers the build and package steps

Reviewer 4: ```suggestion
5. Run the branch through the "hot-fix" build pipeline on Jenkins that *just* executes the build and package steps
```

6. The build artifact can be released to the specific target platform as version `m.n.o.<sha>`

> 7. Open PR to merge your the hot-fix branch into the `master` and `rel_m_n` branches, as per the standard Patch Release process

Reviewer 4: ```suggestion
7. Open PR to merge the hot-fix branch into the `master` and `rel_m_n` branches, as per the standard Patch Release process
```
